### PR TITLE
Parse project information from the CMC project page

### PR DESCRIPTION
**Pull request by valo**
Parse various information about a given project from the CMC page.

### CODE REVIEW

File: lib/sanbase/external_services/coinmarketcap/scraper.ex [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  def fetch_project_page(coinmarketcap_id) do
+    %Tesla.Env{status: 200, body: body} = get("/#{coinmarketcap_id}/")
+
+    body
+  end
+
+  def parse_project_page(html, project_info) do
+    project_info
+    |> struct!(
```
Comment (IvanIvanoff) [inline lib/sanbase/external_services/coinmarketcap/scraper.ex]: What is the reason to use `struct!` instead of `%ProjectInfo{name: name(html) ...}` ?

Comment (valo) [reply lib/sanbase/external_services/coinmarketcap/scraper.ex]: The idea is to have a `ProjectInfo` structure, which is passed through all the methods and "enriched" with more info at each step.

Comment (IvanIvanoff) [reply lib/sanbase/external_services/coinmarketcap/scraper.ex]: That is also equivalent to `%ProjectInfo{project_info | name: name ...}`, updating only the name. This way the code is more explicit about the types, which in my opinion is a desirable result.

File: lib/sanbase/model/project.ex [hunk 2]
```diff
@@ -37,4 +38,12 @@ defmodule Sanbase.Model.Project do
     |> validate_required([:name])
     |> unique_constraint(:name)
   end
+
+  def initial_ico(%Project{id: id}) do
+    Ico
+    |> where([i], i.project_id == ^id)
+    |> order_by(asc: :start_date)
```
Comment (IvanIvanoff) [inline lib/sanbase/model/project.ex]: You can use https://hexdocs.pm/ecto/Ecto.Query.html#first/2

File: lib/sanbase/external_services/coinmarketcap/scraper.ex (hunk 1 above)
Comment (valo) [reply lib/sanbase/external_services/coinmarketcap/scraper.ex]: I agree. I’ll change it

**Comment by IvanIvanoff** [comment]
As this PR is touching coinmarketcap.ex I think it's a good time to move it inside external_services/coinmarketcap directory

**Comment by valo** [comment]
Actually I think the pattern to have `Sanbase.Coinmarketcap` module, which is used as a facade in front of many `Fanbase.Coinmarketcap.*` modules is a nice one. This way we can split functionality and know that the main module is a kind of a public interface.

File: lib/sanbase/external_services/coinmarketcap.ex [hunk 3]
```diff
@@ -70,6 +73,38 @@ defmodule Sanbase.ExternalServices.Coinmarketcap do
     Application.get_env(:sanbase, __MODULE__)
   end
 
+  defp fetch_project_data(project) do
+    fetch_project_info(project)
```
Comment (IvanIvanoff) [inline lib/sanbase/external_services/coinmarketcap.ex]: Let's keep it this way for now, but I think it is not necessary to try to fetch project info every 5 minutes (the default CMC update interval). Could be reworked by starting two types of tasks - :sync_prices and :sync_project_info, which are handled by different `handle_cast`s

Comment (valo) [reply lib/sanbase/external_services/coinmarketcap.ex]: The idea is that if there is no missing data, the project info won't be updated.

**Pull request merged**